File: 2021/day/3/main.py
import fileinput

with fileinput.input() as f:
    N = 12
    numbers = [ln.strip() for ln in f]
    print(numbers)

    # Part 1
    gamma = 0
    epsilon = 0
    n0 = [0] * N
    n1 = [0] * N

    for i in range(N):
        for x in numbers:
            if x[i] == '0':
                n0[i] += 1
            elif x[i] == '1':
                n1[i] += 1

        if n0[i] < n1[i]:
            epsilon += 1 << (N - i - 1)
        else:
            gamma += 1 << (N - i - 1)

    print(gamma * epsilon)
        
    oxygen = set(numbers)
    
    for i in range(N):
        n0 = 0
        n1 = 0

        for s in oxygen:
            if s[i] == '0':
                n0 += 1
            elif s[i] == '1':
                n1 += 1

        if n0 > n1:
            oxygen = {s for s in oxygen if s[i] == '0'}
        else:
            oxygen = {s for s in oxygen if s[i] == '1'}

        if len(oxygen) == 1:
            break

    assert len(oxygen) == 1

    # co2
    co2 = set(numbers)

    for i in range(N):
        n0 = 0
        n1 = 0

        for s in co2:
            if s[i] == '0':
                n0 += 1
            elif s[i] == '1':
                n1 += 1

        if n0 <= n1:
            co2 = {s for s in co2 if s[i] == '0'}
        else:
            co2 = {s for s in co2 if s[i] == '1'}

        if len(co2) == 1:
            break

    assert len(co2) == 1

    ans2 = next(int(n, 2) for n in oxygen) * \
           next(int(n, 2) for n in co2)
    print(ans2)

File: 2021/day/3/Main.hs
import Control.Arrow ((***))
import Data.Char (digitToInt)
import Data.List
import Data.Monoid

toDecimal :: String -> Int
toDecimal = foldl' (\acc x -> acc * 2 + digitToInt x) 0

countDigits :: [String] -> [(Int, Int)]
countDigits numbers = map ((length *** length) . partition (== '0')) (transpose numbers)

gammaRate = map f . countDigits
  where f (n0, n1) | n0 > n1   = '0'
                   | otherwise = '1'

diagnosticReport :: (Int -> Int -> Bool) -> [String] -> String
diagnosticReport bitCriteria numbers = head $ snd $ until done sel (0, numbers)
  where
    done (_, [x]) = True  -- Found it
    done _        = False

    sel (i, xs) =
        let (xs0, xs1) = partition ((== '0') . (!! i)) xs
        in if bitCriteria (length xs0) (length xs1)
               then (i + 1, xs1)
               else (i + 1, xs0)

main = do
    input <- readFile "input.txt"
    let numbers = lines input

    -- Part 1
    let gamma = gammaRate numbers
        epsilon = map f gamma  -- Can simply invert the gamma rate
          where f '0' = '1'
                f '1' = '0'
    print $ toDecimal gamma * toDecimal epsilon

    -- Part 2
    let oxygen = diagnosticReport (<=) numbers
        co2 = diagnosticReport (>) numbers
    print $ toDecimal oxygen * toDecimal co2
